routing/missioncontrol_store.go

```diff
@@ -0,0 +1,235 @@
+		resultBucket := tx.Bucket(resultsKey)
+		results = make([]*paymentResult, resultBucket.Stats().KeyN)
+
+		i := 0
+		return resultBucket.ForEach(func(k, v []byte) error {
+			result, err := deserializeResult(k, v)
+			if err != nil {
+				return err
+			}
+
+			results[i] = result
+			i++
+
+			return nil
+		})
+
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return results, nil
+}
+
+// deserializeResult deserializes a payment result.
+func deserializeResult(k, v []byte) (
+	*paymentResult, error) {
+
+	// Parse payment id.
+	result := paymentResult{
+		id: byteOrder.Uint64(k[8:]),
+	}
+
+	r := bytes.NewReader(v)
+
+	// Read timestamps, success status and failure source index.
+	var (
+		timeFwd, timeReply uint64
+		dbFailureSourceIdx int32
+	)
+
+	err := channeldb.ReadElements(
+		r, &timeFwd, &timeReply, &result.success, &dbFailureSourceIdx,
+	)
+	if err != nil {
+		return nil, err
+	}
+	result.timeFwd = time.Unix(0, int64(timeFwd)).UTC()
+	result.timeReply = time.Unix(0, int64(timeReply)).UTC()
+	if dbFailureSourceIdx != -1 {
+		failureSourceIdx := int(dbFailureSourceIdx)
+		result.failureSourceIdx = &failureSourceIdx
+	}
+
+	// Read route.
+	route, err := channeldb.DeserializeRoute(r)
+	if err != nil {
+		return nil, err
+	}
+	result.route = &route
+
+	// Read failure.
+	failureBytes, err := wire.ReadVarBytes(
+		r, 0, lnwire.FailureMessageLength, "failure",
+	)
+	if err != nil {
+		return nil, err
+	}
+	if len(failureBytes) > 0 {
+		result.failure, err = lnwire.DecodeFailureMessage(
+			bytes.NewReader(failureBytes), 0,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &result, nil
+}
+
+// AddResult adds a new result to the db.
+func (b *missionControlStore) AddResult(rp *paymentResult) error {
+	return b.db.Update(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(resultsKey)
+
+		// Prune oldest entries.
+		if b.maxRecords > 0 {
+			count := bucket.Stats().KeyN
+			for count >= b.maxRecords {
+				cursor := bucket.Cursor()
+				cursor.First()
+				if err := cursor.Delete(); err != nil {
+					return err
+				}
+
+				count--
+			}
+		}
+
+		// Write timestamps, success status, failure source index and
+		// route.
+		var b bytes.Buffer
+
+		var dbFailureSourceIdx int32
+		if rp.failureSourceIdx == nil {
+			dbFailureSourceIdx = -1
+		} else {
+			dbFailureSourceIdx = int32(*rp.failureSourceIdx)
+		}
+
+		err := channeldb.WriteElements(
+			&b,
+			uint64(rp.timeFwd.UnixNano()),
+			uint64(rp.timeReply.UnixNano()),
+			rp.success, dbFailureSourceIdx,
+		)
+		if err != nil {
+			return err
+		}
+
+		if err := channeldb.SerializeRoute(&b, *rp.route); err != nil {
+			return err
+		}
+
+		// Write failure. If there is no failure message, write an empty
+		// byte slice.
+		var failureBytes bytes.Buffer
+		if rp.failure != nil {
+			err := lnwire.EncodeFailureMessage(&failureBytes, rp.failure, 0)
+			if err != nil {
+				return err
+			}
+		}
+		err = wire.WriteVarBytes(&b, 0, failureBytes.Bytes())
```

Comment:
definitely not a blocker, but could be nice to have an equivalent `serializeResult` method so that's it's easier to examine to encoding and decoding logic side-by-side

Reply:
added